File: enso/src/core/system/boot/grammar_schema.rb

require 'schema/schemagen'
require 'schema/schemaschema'

class GrammarSchema < SchemaGenerator
  primitive :str
  primitive :int
  primitive :bool

  def self.print_paths
    { :rules => {} }
  end
  
  klass Grammar do
    field :name, :type => :str
    field :start, :type => Rule
    field :rules, :type => Rule, :optional => true, :many => true
  end

  klass Rule do
    field :name, :type => :str, :key => true
    field :grammar, :type => Grammar, :inverse => Grammar.rules, :key => true
    field :arg, :type => Expression
  end

  klass Expression do
  end
    
  klass Alt, :super => Expression do
    field :alts, :type => Expression, :many => true
  end

  klass Sequence, :super => Expression do
    field :elements, :type => Expression, :optional => true, :many => true
  end

  klass Create, :super => Expression do
    field :name, :type => :str
    field :arg, :type => Expression
  end

  klass Field, :super => Expression do
    field :name, :type => :str
    field :arg, :type => Expression
  end
  
  klass Code, :super => Expression do
    field :code, :type => :str
  end

  klass Value, :super => Expression do
    field :kind, :type => :str
  end

  klass Ref, :super => Expression do
    field :name, :type => :str
  end

  klass Lit, :super => Expression do
    field :value, :type => :str
  end

  klass Call, :super => Expression do 
    field :rule, :type => Rule
  end

  klass Regular, :super => Expression do
    field :arg, :type => Expression
    field :optional, :type => :bool
    field :many, :type => :bool
    field :sep, :type => :str, :optional => true
  end

  SchemaSchema.finalize(schema)
end

if __FILE__ == $0 then

  require 'schema/schemaschema'
  require 'tools/print'
  
  Print.new.recurse(GrammarSchema.schema, SchemaSchema.print_paths)  
end

File: src/core/lang/kotlin/grammar_schema.kt
package Grammar
import schema.Many
interface Grammar {
  var start : Rule
  val rules : Map<String, Rule>
}
interface Pattern
interface Rule {
  var name : String
  var grammar : Grammar
  var arg : Pattern?
}
interface Alt : Pattern {
  val alts : Many<Pattern>
}
interface Sequence : Pattern {
  val elements : Many<Pattern>
}
interface Create : Pattern {
  var name : String
  var arg : Pattern
}
interface Field : Pattern {
  var name : String
  var arg : Pattern
}
interface Terminal : Pattern
interface Value : Terminal {
  var kind : String
}
interface Ref : Terminal {
  var path : Expr
}
interface Lit : Terminal {
  var value : String
}
interface Call : Pattern {
  var rule : Rule
}
interface Regular : Pattern {
  var arg : Pattern
  var optional : Boolean
  var many : Boolean
  var sep : Pattern?
}
interface NoSpace : Pattern
interface Break : Pattern {
  var lines : Int
}
interface Indent : Pattern {
  var indent : Int
}
interface Hide : Pattern {
  var arg : Pattern
}
interface Code : Terminal {
  var expr : Expr
}
interface Expr
interface ETernOp : Expr {
  var op1 : String
  var op2 : String
  var e1 : Expr
  var e2 : Expr
  var e3 : Expr
}
interface EBinOp : Expr {
  var op : String
  var e1 : Expr
  var e2 : Expr
}
interface EUnOp : Expr {
  var op : String
  var e : Expr
}
interface InstanceOf : Expr {
  var base : Expr
  var class_name : String
}
interface EFunCall : Expr {
  var function : Expr
  val params : Many<Expr>
  var lambda : ELambda?
}
interface ELambda {
  var body : Expr
  val formals : Many<Formal>
}
interface Formal {
  var name : String
}
interface EField : Expr {
  var e : Expr
  var fname : String
}
interface EVar : Expr {
  var name : String
}
interface ESubscript : Expr {
  var e : Expr
  var sub : Expr
}
interface EList : Expr {
  val elems : Many<Expr>
}
interface ENew : Expr {
  var cls : String
}
interface EConst : Expr
interface EStrConst : EConst {
  var value : String
  val type : String
}
interface EIntConst : EConst {
  var value : Int
  val type : String
}
interface EBoolConst : EConst {
  var value : Boolean
  val type : String
}
interface ERealConst : EConst {
  var value : Float
  val type : String
}
interface ENil : EConst
